src/darwin/Framework/CHIP/MTRCertificateInfo.mm

```diff
@@ -0,0 +1,185 @@
+}
+
+- (nullable instancetype)initWithTLVBytes:(MTRCertificateTLVBytes)bytes
+{
+    if (self = [super init]) {
+        VerifyOrReturnValue(_bytes = [bytes copy], nil);
+        VerifyOrReturnValue(DecodeChipCert(AsByteSpan(_bytes), _data) == CHIP_NO_ERROR, nil);
+    }
+    return self;
+}
+
+- (MTRDistinguishedNameInfo *)issuer
+{
+    return [[MTRDistinguishedNameInfo alloc] initWithDN:_data.mIssuerDN];
+}
+
+- (MTRDistinguishedNameInfo *)subject
+{
+    return [[MTRDistinguishedNameInfo alloc] initWithDN:_data.mSubjectDN];
+}
+
+- (NSDate *)notBefore
+{
+    return ChipEpochSecondsAsDate(_data.mNotBeforeTime);
+}
+
+- (NSDate *)notAfter
+{
+    // "no expiry" is encoded as kNullCertTime (see ChipEpochToASN1Time)
+    return (_data.mNotAfterTime != kNullCertTime) ? ChipEpochSecondsAsDate(_data.mNotAfterTime) : NSDate.distantFuture;
+}
+
+- (id)copyWithZone:(nullable NSZone *)zone
+{
+    return self;
+}
+
+- (BOOL)isEqual:(id)object
+{
+    if (self == object) {
+        return YES;
+    }
+    if ([object class] != MTRCertificateInfo.class) {
+        return NO;
+    }
+    auto other = (MTRCertificateInfo *) object;
+    return [_bytes isEqual:other->_bytes];
+}
+
+- (NSUInteger)hash
+{
+    return _bytes.hash;
+}
+
+@end
+
+MTR_DIRECT_MEMBERS
+@implementation MTRDistinguishedNameInfo {
+    ChipDN _dn;
+}
+
+- (instancetype)initWithDN:(const ChipDN &)dn
+{
+    if (self = [super init]) {
+        _dn = dn;
+    }
+    return self;
+}
+
+- (nullable NSNumber *)nodeID
+{
+    return [self identifierWithUniqueOID:kOID_AttributeType_MatterNodeId];
+}
+
+- (nullable NSNumber *)fabricID
+{
+    return [self identifierWithUniqueOID:kOID_AttributeType_MatterFabricId];
+}
+
+- (nullable NSNumber *)rootCACertificateID
+{
+    return [self identifierWithUniqueOID:kOID_AttributeType_MatterRCACId];
+}
+
+- (nullable NSNumber *)intermediateCACertificateID
+{
+    return [self identifierWithUniqueOID:kOID_AttributeType_MatterICACId];
+}
+
+- (nullable NSNumber *)identifierWithUniqueOID:(OID)oid
+{
+    NSAssert(IsChipDNAttr(oid), @"Invalid OID");
+
+    ChipRDN const * match = nullptr;
+    for (auto const & rdn : _dn.rdn) {
+        if (rdn.IsEmpty()) {
+            break;
+        } else if (rdn.mAttrOID == oid) {
+            VerifyOrReturnValue(!match, nil); // invalid, there should be only one matching RDN
+            match = &rdn;
+        }
+    }
+
+    return (match) ? @(match->mChipVal) : nil;
+}
+
+- (NSSet<NSNumber *> *)tags
+{
+    NSMutableSet<NSNumber *> * result;
+    for (auto const & rdn : _dn.rdn) {
+        if (rdn.IsEmpty()) {
+            break;
+        } else if (rdn.mAttrOID == kOID_AttributeType_MatterCASEAuthTag) {
+            NSNumber * tag = @(rdn.mChipVal);
+            if (!result) {
+                result = [NSMutableSet setWithObject:tag];
+            } else {
+                [result addObject:tag];
+            }
+        }
+    }
+    return (result) ? [result copy] : [NSSet set];
```

Comment:
Why do we need a copy here instead of just returning `result`?

Reply:
It's not strictly necessary, just a defensive programming habit.